### ISSUE TITLE
ZManaged inside ZStream release resources too early in parallel mode (mapMPar) 

### ISSUE THREAD

**Issue by heaven-born** [zio/zio#4102]
There is a suspicions behaviour of ZStream combined with ZManaged.
 
    val runtime = zio.Runtime.default
    val foo = Managed.make(console.putStrLn("capturing foo") *> ZIO("foo"))(_ => console.putStrLn("released foo"))
    val bar = Managed.make(console.putStrLn("capturing bar") *> ZIO("bar"))(_ => console.putStrLn("released bar"))

    val k = ZStream(foo) ++ ZStream(bar)
    val r = k.flatMap(s => ZStream.managed(s))
                 .mapMPar(3)(x => console.putStrLn(s"Processing `${x}`"))

    runtime.unsafeRun(r.runDrain)

   The code above returns the following 

    capturing foo
    released foo
    Processing `foo`
    capturing bar
    Processing `bar`
    released bar

which is incorrect I assume, because resource "foo" is released before it actually used inside "mapMPar".
The issue is not reproducible on regular "mapM".

zio/zio-streams v1.0.0/scala 2.13.0

**Comment by iravid** [comment]
If you'd like to maintain a resource for the lifetime of a stream, the stream that uses the resource has to be enclosed in the scope of a flatMap on the stream producing the resource.

In your example, the only way this could work is by writing:
```scala
(ZStream(foo) ++ ZStream(bar))
  .flatMap { s =>
    ZStream.managed(s)
  }
  .flatMap { resource =>
    ZStream(resource) // or produce another stream, somehow, from resource
      .mapMPar(3)(x => console.putStrLn(...)
  }
```

So although surprising, this is not a bug. The behavior that occurs in your example is:
1. `foo` is pulled from `k`
2. `foo` is transformed to a stream of one element that maintains the ZManaged scope
3. the value produced by the ZManaged is pulled from the stream in 2 by `mapMPar`, which causes the stream to end and release the resource
4. 1-3 happen concurrently with mapMPar's processing, so there's no guarantee here on the order of finalization.

**Comment by heaven-born** [comment]
Now I see. But wouldn't be more logical to delegate resource maintaining responsibility to the new enclosing stream on flatMap?

And one more strange behaviour in my example: mapMPar is not actually doing any parallelization. "foo" is always processed before "bar" even if  to slow down "foo" using ZIO.sleep .

**Comment by heaven-born** [comment]
The more I think about it, the more I find that current implementation is more natural. The merging method that I mentioned may help in some cases I guess, but I can't find good practical application to it.

**Comment by heaven-born** [comment]
Closing this issue as "not a bug". If I find that the second problem mentioned in this ticket is a bug, I will open new ticket.
